### PR TITLE
Add a note about nested method declarations

### LINKED ISSUE #437: Add a note about nested function declaration
Unfortunately sometimes I see something like this
```ruby
def foo(x)
  def bar(y)
    do_work y
  end
  bar(x)
end
```
in the wild and it's author doesn't know that it isn't nested function declaration, but full-blown function declaration (and re-declaration on each foo() call).

I think that it won't hurt if there will be mention of this with suggestion to use lambda/Proc or "non-nested" function declaration.

### PR DESCRIPTION
**Pull request by ojab**
Fixes #437

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1421,6 +1421,36 @@ condition](#safe-assignment-in-condition).
   you forget either of the rules above!
 <sup>[[link](#always-warn-at-runtime)]</sup>
 
+* <a name="no-nested-methods"></a>
+  Do not use nested method declarations, use lambda instead.
```
Comment (bbatsov) [inline README.md]: declarations -> definitions

Comment (ojab) [inline README.md]: fxd

File: README.md [hunk 2]
```diff
@@ -1421,6 +1421,36 @@ condition](#safe-assignment-in-condition).
   you forget either of the rules above!
 <sup>[[link](#always-warn-at-runtime)]</sup>
 
+* <a name="no-nested-methods"></a>
+  Do not use nested method definitions, use lambda instead.
+  Nested method definition is equal to the definition in the same scope as
```
Comment (bbatsov) [inline README.md]: Nested method definitions actually produce methods in the same scope (e.g. class) as the out method. Furthermore, the "nested method" will be redefined every time the method containing its definition is invoked.

File: README.md [hunk 3]
```diff
@@ -1421,6 +1421,36 @@ condition](#safe-assignment-in-condition).
...
+  the outer method and [re]defining inner method on every outer method call.
+<sup>[[link](#no-nested-methods)]</sup>
+
+  ```Ruby
```
Comment (bbatsov) [inline README.md]: This code block is not indented as the blocks in other code examples.

File: README.md [hunk 4]
```diff
@@ -1421,6 +1421,36 @@ condition](#safe-assignment-in-condition).
...
+  # bad
+  def foo(x)
+    def bar(y)
+      # body omitted
+    end
+    bar(x)
```
Comment (bbatsov) [inline README.md]: I'd add a blank line before `bar(x)`.

File: README.md [hunk 5]
```diff
@@ -1421,6 +1421,36 @@ condition](#safe-assignment-in-condition).
...
+  end
+
+  # good, the same as the previous, but no bar redefinition on every foo call
```
Comment (bbatsov) [inline README.md]: good -

File: README.md [hunk 6]
```diff
@@ -1421,6 +1421,36 @@ condition](#safe-assignment-in-condition).
...
+  def bar(y)
+    # body omitted
+  end
+  def foo(x)
```
Comment (bbatsov) [inline README.md]: Add a blank line between the defs.

File: README.md [hunk 7]
```diff
@@ -1421,6 +1421,40 @@ condition](#safe-assignment-in-condition).
...
+    # body omitted
+  end
+  
+  def foo(x)
+    bar(x)
+  end
+
+  # also good
+  def foo(x)
+    bar = ->(y) { # body omitted }
```
Comment (abotalov) [inline README.md]: `}` will also be shown greyed out on Github.

File: README.md [hunk 8]
```diff
@@ -1421,6 +1421,40 @@ condition](#safe-assignment-in-condition).
   you forget either of the rules above!
 <sup>[[link](#always-warn-at-runtime)]</sup>
 
+* <a name="no-nested-methods"></a>
+  Do not use nested method definitions, use lambda instead.
+  Nested method definitions actually produce methods in the same scope
+  (e.g. class) as the out method. Furthermore, the "nested method" will be
+  redefined every time the method containing its definition is invoked.
+
+<sup>[[link](#no-nested-methods)]</sup>
```
Comment (abotalov) [inline README.md]: `[link]` shouldn't be shown on the separate line, I believe.

File: README.md (hunk 7 above)
Comment (ojab) [inline README.md]: Replaced it with { ... } as in [reverse-each](https://github.com/bbatsov/ruby-style-guide#reverse-each) suggestion.

File: README.md (hunk 8 above)
Comment (ojab) [inline README.md]: fixed, thanks

File: README.md [hunk 9]
```diff
@@ -1421,6 +1421,40 @@ condition](#safe-assignment-in-condition).
   you forget either of the rules above!
 <sup>[[link](#always-warn-at-runtime)]</sup>
 
+* <a name="no-nested-methods"></a>
+  Do not use nested method definitions, use lambda instead.
+  Nested method definitions actually produce methods in the same scope
+  (e.g. class) as the out method. Furthermore, the "nested method" will be
```
Comment (bbatsov) [inline README.md]: out -> outer

Comment (ojab) [inline README.md]: fxd

**Comment by bbatsov** [comment]
:+1: Thanks!

**Comment by ojab** [comment]
Actually thank *you*.
I will learn proper English one day, heh.

**Pull request merged**